jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt

```diff
@@ -0,0 +1,305 @@
+ * This app is hosted in github repo - https://github.com/JetBrains/JetBrains.Rider.Debugger.Launcher and is
+ * delivered via GutHub releases with packed output.
+ * This output is copied to the same Rider SDK path that is mounted in Docker (app should be placed under "Amazon.Runner" folder).
+ *
+ * After Debugger Worker process is started on Docker instance it initializes ports and process to run - MockBootstraps.dll
+ * (the internal compiled with lambda code)
+ */
+class DotNetSamDebugSupport : SamDebugSupport {
+    companion object {
+        private val logger = getLogger<DotNetSamDebugSupport>()
+
+        private const val debuggerDir = "/tmp/lambci_debug_files"
+        private const val dotnetCorCliPath = "/var/lang/bin/dotnet"
+        private const val mockBootstrapsPath = "/var/runtime/MockBootstraps.dll"
+    }
+
+    /**
+     * Check whether the Rider Debugger Worker Launcher app (that is required to run Debugger) is downloaded into Rider SDK.
+     */
+    override fun isSupported(): Boolean {
+        val debuggerWorkerAssembly = RiderEnvironment.getBundledFile(DebuggerWorkerPlatform.AnyCpu.assemblyName)
+        val binDir = debuggerWorkerAssembly.parentFile
+        val debugLauncherFile = File(binDir, "Amazon.Runner").resolve("JetBrains.Debugger.Launcher.dll")
+
+        val debuggerLauncherExists = debugLauncherFile.exists()
+        if (!debuggerLauncherExists) {
+            logger.error { "JetBrains.Rider.Debugger.Launcher runnable does not exists" }
+        }
+        return debuggerLauncherExists
+    }
+
+    override fun patchCommandLine(debugPort: Int, state: SamRunningState, commandLine: GeneralCommandLine) {
+        val debuggerWorkerAssembly = RiderEnvironment.getBundledFile(DebuggerWorkerPlatform.AnyCpu.assemblyName)
+        val binDir = debuggerWorkerAssembly.parentFile
+        val debuggerMode = "server"
+
+        val debugArgs = "$debuggerDir/Amazon.Runner/JetBrains.Debugger.Launcher.dll $debuggerDir/runtime.sh $debuggerDir/${debuggerWorkerAssembly.name} $debuggerMode $debugPort $debugPort"
+
+        commandLine.withParameters("--debugger-path")
+                .withParameters(binDir.path)
+                .withParameters("--debug-args")
+                .withParameters(debugArgs)
+
+        super.patchCommandLine(debugPort, state, commandLine)
+    }
+
+    override fun createDebugProcess(
+        environment: ExecutionEnvironment,
+        state: SamRunningState,
+        debugPort: Int
+    ): XDebugProcessStarter? {
+        throw UnsupportedOperationException("Use 'createDebugProcessAsync' instead")
+    }
+
+    override fun createDebugProcessAsync(
+        environment: ExecutionEnvironment,
+        state: SamRunningState,
+        debugPort: Int
+    ): Promise<XDebugProcessStarter?> {
+
+        val promise = AsyncPromise<XDebugProcessStarter?>()
+
+        val project = environment.project
+        val debuggerLifetimeDefinition = project.defineNestedLifetime()
```

Comment:
Can you write a bit of a comment on what these do since I don't think we will have the javadoc

Reply:
Added a simple commend about debuggerLifetime. Very generally, lifetime is a base concept that we use in Rider to control disposable objects within container layers. Please let me know if I should extend this a bit.